File: Trees/Binary Trees/inorder_traversal.js
class TreeNode {
  constructor(val) {
    this.val = val;
    this.left = this.right = null;
  }
}

class Solution {
  inorderTraversal(root) {
    const result = [];
    const stack = [];
    let current = root;

    while (current || stack.length > 0) {
      // Traverse all the way to the leftmost node, pushing each node onto the stack
      while (current) {
        stack.push(current);
        current = current.left;
      }

      // Pop the top node from the stack (current leftmost node)
      current = stack.pop();

      // Add the value of the current node to the result array
      result.push(current.val);

      // Move to the right subtree of the current node
      current = current.right;
    }

    return result;
  }
}

// Example usage
const root = new TreeNode(1);
root.right = new TreeNode(2);
root.right.left = new TreeNode(3);

const solution = new Solution();
const result = solution.inorderTraversal(root);

console.log(result); // Output: [1, 3, 2]

File: Trees/Binary Trees/inorder_traversal.go
package main

import "fmt"

// TreeNode definition
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
	var result []int
	var stack []*TreeNode
	current := root

	for current != nil || len(stack) > 0 {
		// Traverse all the way to the leftmost node, pushing each node onto the stack
		for current != nil {
			stack = append(stack, current)
			current = current.Left
		}

		// Pop the top node from the stack (current leftmost node)
		current, stack = stack[len(stack)-1], stack[:len(stack)-1]

		// Add the value of the current node to the result slice
		result = append(result, current.Val)

		// Move to the right subtree of the current node
		current = current.Right
	}

	return result
}

func main() {
	// Create a sample binary tree
	root := &TreeNode{Val: 1, Right: &TreeNode{Val: 2, Left: &TreeNode{Val: 3}}}

	// Perform in-order traversal
	result := inorderTraversal(root)

	// Print the result
	fmt.Println(result) // Output: [1 3 2]
}
